modules/tinymce/src/core/main/ts/content/InsertContentImpl.ts

```diff
@@ -0,0 +1,335 @@
+import { CaretWalker } from '../caret/CaretWalker';
+import ElementUtils from '../api/dom/ElementUtils';
+import * as NodeType from '../dom/NodeType';
+import * as PaddingBr from '../dom/PaddingBr';
+import Serializer from '../api/html/Serializer';
+import * as RangeNormalizer from '../selection/RangeNormalizer';
+import Selection from '../api/dom/Selection';
+import Editor from '../api/Editor';
+import DOMUtils from '../api/dom/DOMUtils';
+import { trimOrPadLeftRight, trimNbspAfterDeleteAndPadValue, isAfterNbsp } from './NbspTrim';
+import ParserNode from '../api/html/Node';
+import Tools from '../api/util/Tools';
+
+const isTableCell = NodeType.matchNodeNames([ 'td', 'th' ]);
+
+const selectionSetContent = (editor: Editor, content: string) => {
+  const rng = editor.selection.getRng();
+  const container = rng.startContainer;
+  const offset = rng.startOffset;
+
+  if (rng.collapsed && isAfterNbsp(container, offset) && NodeType.isText(container)) {
+    container.insertData(offset - 1, ' ');
+    container.deleteData(offset, 1);
+    rng.setStart(container, offset);
+    rng.setEnd(container, offset);
+    editor.selection.setRng(rng);
+  }
+
+  editor.selection.setContent(content);
+};
+
+const validInsertion = function (editor: Editor, value: string, parentNode: DomElement) {
+  // Should never insert content into bogus elements, since these can
+  // be resize handles or similar
+  if (parentNode.getAttribute('data-mce-bogus') === 'all') {
+    parentNode.parentNode.insertBefore(editor.dom.createFragment(value), parentNode);
+  } else {
+    // Check if parent is empty or only has one BR element then set the innerHTML of that parent
+    const node = parentNode.firstChild;
+    const node2 = parentNode.lastChild;
+    if (!node || (node === node2 && node.nodeName === 'BR')) {
+      editor.dom.setHTML(parentNode, value);
+    } else {
+      selectionSetContent(editor, value);
+    }
+  }
+};
+
+const trimBrsFromTableCell = function (dom: DOMUtils, elm: DomElement) {
+  Option.from(dom.getParent(elm, 'td,th')).map(Element.fromDom).each(PaddingBr.trimBlockTrailingBr);
+};
+
+const reduceInlineTextElements = (editor: Editor, merge: boolean) => {
+  const textInlineElements = editor.schema.getTextInlineElements();
+  const dom = editor.dom;
+
+  if (merge) {
+    const root = editor.getBody(), elementUtils = new ElementUtils(dom);
+
+    Tools.each(dom.select('*[data-mce-fragment]'), function (node) {
+      for (let testNode = node.parentNode; testNode && testNode !== root; testNode = testNode.parentNode) {
+        if (textInlineElements[node.nodeName.toLowerCase()] && elementUtils.compare(testNode, node)) {
+          dom.remove(node, true);
+        }
+      }
+    });
+  }
+};
+
+const markFragmentElements = (fragment: ParserNode) => {
+  let node = fragment;
+
+  while ((node = node.walk())) {
+    if (node.type === 1) {
+      node.attr('data-mce-fragment', '1');
+    }
+  }
+};
+
+const umarkFragmentElements = (elm: DomElement) => {
```

Comment:
unmark?

Reply:
literally just extracted from source, I tried not to change anything I didn't need to
https://github.com/tinymce/tinymce/blob/develop/modules/tinymce/src/core/main/ts/content/InsertContent.ts#L93

Reply:
I can go in an change this before I merge.